idle.go

```diff
@@ -0,0 +1,390 @@
+// operations to synchronize access to shared state and a mutex to guarantee
+// mutual exclusion in a critical section.
+type atomicIdlenessManager struct {
+	// State accessed atomically.
+	lastCallEndTime           int64 // Unix timestamp in nanos; time when the most recent RPC completed.
+	activeCallsCount          int32 // Count of active RPCs; -math.MaxInt32 means channel is idle or is trying to get there.
+	activeSinceLastTimerCheck int32 // Boolean; True if there was an RPC since the last timer callback.
+	closed                    int32 // Boolean; True when the manager is closed.
+
+	// Can be accessed without atomics or mutex since these are set at creation
+	// time and read-only after that.
+	enforcer idlenessEnforcer // Functionality provided by grpc.ClientConn.
+	timeout  int64            // Idle timeout duration nanos stored as an int64.
+
+	// idleMu is used to guarantee mutual exclusion in two scenarios:
+	// - Opposing intentions:
+	//   - a: Idle timeout has fired and handleIdleTimeout() is trying to put
+	//     the channel in idle mode because the channel has been inactive.
+	//   - b: At the same time an RPC is made on the channel, and onCallBegin()
+	//     is trying to prevent the channel from going idle.
+	// - Competing intentions:
+	//   - The channel is in idle mode and there are multiple RPCs starting at
+	//     the same time, all trying to move the channel out of idle. Only one
+	//     of them should succeed in doing so, while the other RPCs should
+	//     piggyback on the first one and be successfully handled.
+	idleMu       sync.RWMutex
+	actuallyIdle bool
+	timer        *time.Timer
+}
+
+// newAtomicIdlenessManager creates a new atomicIdlenessManager. A non-zero
+// value must be passed for idle timeout.
+func newAtomicIdlenessManager(enforcer idlenessEnforcer, idleTimeout time.Duration) idlenessManager {
+	i := &atomicIdlenessManager{
+		enforcer: enforcer,
+		timeout:  int64(idleTimeout),
+	}
+	i.timer = timeAfterFunc(idleTimeout, i.handleIdleTimeout)
+	return i
+}
+
+// resetIdleTimer resets the idle timer to the given duration. This method
+// should only be called from the timer callback.
+func (i *atomicIdlenessManager) resetIdleTimer(d time.Duration) {
+	i.idleMu.Lock()
+	defer i.idleMu.Unlock()
+
+	if i.timer == nil {
+		// Only close sets timer to nil. We are done.
+		return
+	}
+
+	// It is safe to ignore the return value from Reset() because this method is
+	// only ever called from the timer callback, which means the timer has
+	// already fired.
+	i.timer.Reset(d)
+}
+
+// handleIdleTimeout is the timer callback that is invoked upon expiry of the
+// configured idle timeout. The channel is considered inactive if there are no
+// ongoing calls and no RPC activity since the last time the timer fired.
+func (i *atomicIdlenessManager) handleIdleTimeout() {
+	if i.isClosed() {
+		return
+	}
+
+	if atomic.LoadInt32(&i.activeCallsCount) > 0 {
+		i.resetIdleTimer(time.Duration(i.timeout))
+		return
+	}
+
```

Comment:
Move this to within the `atomic.Load` below?

Reply:
Even after our offline discussion I don't quite understand how we can move this to be inside the `if atomic.LoadInt32(&i.activeSinceLastTimerCheck) == 1 { ... }`. If there is already activity since the last timer check, which is when the if condition would be `true`, why do we have to check the `activeCallsCount` at that point?

Reply:
You would need to check it because that's part of the logic of setting the timer.  If you have active RPCs, then you want to set the timer to the maximum idle time.  If you don't have active RPCs, then you set it to the time since the last one completed.  If `activeSinceLastTimerCheck==0` then you probably don't need to set the timer at all (you should go idle) -- unless you lose a race with a brand new RPC.

Reply:
Discussed this offline.

The main concern being the multiple reads of `i.activeCallsCount`. Agreed that its not a big deal and decided to leave it as it for now.